### PR TITLE
Replaced the usage of custom function assert_equal with np.testing.assert_equal

### LINKED ISSUE #15782: Remove needless functions from an `io.ascii` test helper file
### What is the problem this feature will solve?

A helper file in `io.ascii` tests contains three functions that serve little purpose: https://github.com/astropy/astropy/blob/fefe2261fcf09751d052d80bd8143e72f18f6aea/astropy/io/ascii/tests/common.py#L27-L37

### Describe the desired outcome

The aforementioned functions should be removed and any tests still using them should be updated accordingly. Pull requests targeting one or more functions are welcome.

Potential contributors can have a look at #16644 that removed the helper functions from one file and use that as an example.

### Additional context

`assert_equal()` and `assert_true()` should be inlined, but `assert_almost_equal(a, b)` should be replaced with [`np.testing.assert_allclose(a, b)`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) instead of `assert np.allclose(a, b)` because the dedicated testing function produces a better error message if the check fails.

### Performed removals

- [ ] `assert_equal()`
- [ ] `assert_almost_equal()`
- [ ] `assert_true()`

### PR DESCRIPTION
**Pull request by Harsh-pandya**
Change Description:

Performed removals
- [x] assert_equal()
- [x] assert_almost_equal()
- [x] assert_true()

### Description

This pull request is to address ...

Fixes #15782

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by Harsh-pandya** [comment]
Hello @dhomeier, I completed the change mentioned in the issue. But, the following are the errors received in the Git workflow can you please help me with those?

1. [Check if towncrier change log entry is correct](https://github.com/astropy/astropy/actions/runs/7470245754/job/20328627189?pr=15839#logs)


2. Precommit.ci  
    ruff.....................................................................Failed
      - hook id: ruff
      - exit code: 1
      - files were modified by this hook

       astropy/io/ascii/tests/test_c_reader.py:54:25: PT009 Use a regular `assert` instead of unittest-style `assert_`
       
      Over here the function provided by `np.testing is assert_`.  This function is the replacement for the `assert_true
      in common.py.`
      
Apart from this, all the tests ran successfully except 3 failed in my local I have attached a screenshot for those. This failure where in the docs file.
Screenshot 2024-01-09 at 7 06 22 PM

**Review by neutrinoceros — COMMENTED** [review]
Thanks for you contribution !
Just a general comment on style: in astropy, numpy testing function are usually imported one of the following ways
1) `from numpy.testing import assert_array_equal`
2) `import numpy.testing as npt`

**Comment by dhomeier** [comment]
> Just a general comment on style: in astropy, numpy testing function are usually imported one of the following ways
> 
> 1. `from numpy.testing import assert_array_equal`
> 2. `import numpy.testing as npt`

In addition, using the former for `assert_almost_equal` would allow to leave test bodies essentially unmodified.

As for `assert_equal` I am not sure if the intention was to use anything at all but a simple `assert a == b` –
@eerovaher , do the better failure diagnostics you mentioned in #15782 apply here as well (if one does not want to specify a custom error message)?

Last but not least @Rohit-pujari has already begun work on this same task in #15787 – I don't know if there is still a perspective for completing that PR after college classes have resumed.

**Comment by dhomeier** [comment]
> Apart from this, all the tests ran successfully except 3 failed in my local I have attached a screenshot for those. This failure where in the docs file.

The actual detailed errors must be further up in the full test log, but since doc build and tests are passing in the CI, I would not worry too much about that.

**Comment by pllim** [comment]
Gotta do something about that pre-commit rule.

**Comment by dhomeier** [comment]
You mean remove PT009? 🤔
But there were some real fixes of unused or misplaced imports that should be addressed before pushing upstream, @Harsh-pandya . Though I have no idea how the imports would have become unsorted by just removing one – perhaps you need to remove the following blank line as well?

**Comment by pllim** [comment]
Not sure how to properly ignore the rule but I think it should be done locally as needed only. Maybe @eerovaher or @nstarman can advise. Thanks!

**Comment by eerovaher** [comment]
@dhomeier [asked](https://github.com/astropy/astropy/pull/15839#issuecomment-1884983283)

> As for assert_equal I am not sure if the intention was to use anything at all but a simple assert a == b –
@eerovaher , do the better failure diagnostics you mentioned in https://github.com/astropy/astropy/issues/15782 apply here as well (if one does not want to specify a custom error message)?

It's usually best to use the `numpy.testing` helpers for checking array-like collections of numerical values, with the possible exception of checking strict equality, but for anything else I think `pytest` introspection is better. I am not aware of any good reasons to use `np.testing.assert_()`. Have a look at these examples:

Test for lists of numbers:
```python
import numpy as np


values = [1, 2, 3]


def test_example_1():
	assert [1, 3, 3] == values


def test_example_2():
    assert np.array_equal([1, 3, 3], values)


def test_example_3():
    np.testing.assert_array_equal([1, 3, 3], values)


def test_example_4():
	np.testing.assert_([1, 3, 3] == values)
```
Traceback:
```
============================================== FAILURES ===============================================
___________________________________________ test_example_1 ____________________________________________

    def test_example_1():
>   	assert [1, 3, 3] == values
E    assert [1, 3, 3] == [1, 2, 3]
E      At index 1 diff: 3 != 2
E      Use -v to get more diff

test_example.py:8: AssertionError
___________________________________________ test_example_2 ____________________________________________

    def test_example_2():
>       assert np.array_equal([1, 3, 3], values)
E       assert False
E        +  where False = <function array_equal at 0x7f3c07d43eb0>([1, 3, 3], [1, 2, 3])
E        +    where <function array_equal at 0x7f3c07d43eb0> = np.array_equal

test_example.py:12: AssertionError
___________________________________________ test_example_3 ____________________________________________

    def test_example_3():
>       np.testing.assert_array_equal([1, 3, 3], values)

test_example.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<built-in function eq>, [1, 3, 3], [1, 2, 3])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 1 / 3 (33.3%)
E           Max absolute difference: 1
E           Max relative difference: 0.5
E            x: array([1, 3, 3])
E            y: array([1, 2, 3])

/usr/lib/python3.10/contextlib.py:79: AssertionError
___________________________________________ test_example_4 ____________________________________________

    def test_example_4():
>   	np.testing.assert_([1, 3, 3] == values)
E    AssertionError

test_example.py:20: AssertionError
```
Test for strings:
```python
import numpy as np


message = "Expected string output."


def test_example_1():
	assert "Expectd String  output." == message


def test_example_2():
    np.testing.assert_array_equal("Expectd String  output.", message)


def test_example_3():
	np.testing.assert_("Expectd String  output." == message)
```
Traceback:
```
============================================== FAILURES ===============================================
___________________________________________ test_example_1 ____________________________________________

    def test_example_1():
>   	assert "Expectd String  output." == message
E    AssertionError: assert 'Expectd String  output.' == 'Expected string output.'
E      - Expected string output.
E      ?       -  ^
E      + Expectd String  output.
E      ?         ^     +

test_example.py:8: AssertionError
___________________________________________ test_example_2 ____________________________________________

    def test_example_2():
>       np.testing.assert_array_equal("Expectd String  output.", message)

test_example.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<built-in function eq>, 'Expectd String  output.', 'Expected string output.')
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 1 / 1 (100%)
E            x: array('Expectd String  output.', dtype='<U23')
E            y: array('Expected string output.', dtype='<U23')

/usr/lib/python3.10/contextlib.py:79: AssertionError
___________________________________________ test_example_3 ____________________________________________

    def test_example_3():
>   	np.testing.assert_("Expectd String  output." == message)
E    AssertionError

test_example.py:16: AssertionError
```

**Comment by dhomeier** [comment]
> It's usually best to use the `numpy.testing` helpers for checking array-like collections of numerical values, with the possible exception of checking strict equality, but for anything else I think `pytest` introspection is better. I am not aware of any good reasons to use `np.testing.assert_()`. Have a look at these examples:

Thanks; I don't see much benefit in `assert assert np.array_equal()`, but `np.testing.assert_array_equal` certainly offers more useful info especially for large arrays!
I suppose `assert np.testing.assert_` may only become useful if one creates a custom `msg`, but since I don't see a clear case for testing `.pyo` files, I'd leave it at `assert` there (and keep PT008 enabled).

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_c_reader.py:13]:
```suggestion
from contextlib import nullcontext
from io import BytesIO
from textwrap import dedent

import numpy as np
import pytest
from numpy.testing import assert_allclose, assert_equal
```
Frankly I have no idea what import order Ruff wishes to enforce, but the previous form was accepted, so apparently alphabetical order _within_ a block is fine...
I personally prefer the option of using `assert_*` directly; since the old custom functions will be removed entirely, I don't think there is much risk of confusion. But that is up for discussion.

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_c_reader.py:56]:
```suggestion
                        assert not isinstance(t2[name][i], str) and np.isnan(t2[name][i])
```
See discussion; I don't think we want to use `npt.assert_` at all after that.

**Review by dhomeier — CHANGES_REQUESTED** [review]
As discussed I think using just `assert_allclose` and `assert_equal` from `np.testing` will be enough.
Not sure exactly how to make Ruff happy about the import order; perhaps it will be best to let it sort this out itself in the pre-commit hook. So if possible, apply the pre-commit fixes locally on your next commit before pushing. Thanks!

File: astropy/io/ascii/tests/test_c_reader.py
Comment (nstarman) [reply astropy/io/ascii/tests/test_c_reader.py:13]: ruff is configured to the standard `isort` settings: first sort by package origin (standard library, third party, first party, local) then alphabetically within those blocks with `import X` preceding `from X import`.

**Comment by Harsh-pandya** [comment]
@dhomeier  After going through the discussion. I will make the requested changes 👍🏻

**Comment by dhomeier** [comment]
pre-commit.ci autofix

**Comment by dhomeier** [comment]
Anyone has an idea why `autofix` is not doing anything? There also seem to be problems with the local `pre-commit` hook.

**Comment by pllim** [comment]
`push permission was denied -- perhaps the "Allow edits by maintainers" box was unchecked?`

**Comment by dhomeier** [comment]
I don't even see that checkbox anywhere... 🤯

**Comment by pllim** [comment]
Only PR author sees it, I guess.

**Comment by dhomeier** [comment]
Well, have to wait for update from the author then – @Harsh-pandya can you confirm that you have set up and activated `pre-commit` locally as described in the [workflow docs](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit), and what `pre-commit` is reporting as you commit and push?

**Comment by Harsh-pandya** [comment]
> I don't even see that checkbox anywhere... 🤯

@dhomeier  I have enabled it now.

**Comment by Rohit-Pujari** [comment]
Sorry I have my college work so I am unable to work on it

**Comment by Harsh-pandya** [comment]
> Well, have to wait for update from the author then – @Harsh-pandya can you confirm that you have set up and activated `pre-commit` locally as described in the [workflow docs](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit), and what `pre-commit` is reporting as you commit and push (note that I think you have to `git add` and commit again until `pre-commit` is indicating no more corrections)?

@dhomeier  I have enabled the pre-commit in my local and it also fails on ruff for the file `io/tests/test_read.py` I am currently debugging that.

**Comment by dhomeier** [comment]
It seems to complain about the removed blank line 23 before the comment line – no idea why this constitutes an isort issue.
But I still don't understand why it does not then apply the fixes before allowing you to commit.

**Comment by Harsh-pandya** [comment]
@dhomeier All the issues on the pre-commit have been solved. I apologize for not setting up the pre-commit hook locally by reading the docs. I will take care of this next time.  Thanks for your support 👍🏻 .

**Comment by dhomeier** [comment]
> Sorry I have my college work so I am unable to work on it

No problem, as we would rather go with the alternative approach (using the `numpy.testing` functions where possible) here. Thank you for you work so far @Rohit-Pujari, and hope this will give you time to set up a working development installation for your next contribution!

**Review by dhomeier — APPROVED** [review]
Thanks, got everything to pass after all!
I am leaving this open for @taldcroft and @hamogu to comment if they have any preferences on `assert_equal` vs. `npt.assert_equal`; for reference the existing code has 109 uses `from numpy.testing import assert_something` vs. 1 of 
`import numpy.testing as npt`.

Will also need squashing before this can be merged!

**Comment by dhomeier** [comment]
I understood your comparison in https://github.com/astropy/astropy/pull/15839#issuecomment-1885134788

> It's usually best to use the `numpy.testing` helpers for checking array-like collections of numerical values, 

to exactly the opposite effect – `assert_equal` providing more detailed info (in fact, imo, a bit more useful yet than `assert_array_equal` for plain lists etc.):
```
>>> np.testing.assert_equal([1, 3, 3], [1, 2, 3])
AssertionError: 
Items are not equal:
item=1

 ACTUAL: 3
 DESIRED: 2
```

File: astropy/io/ascii/tests/test_c_reader.py
Comment (eerovaher) [inline astropy/io/ascii/tests/test_c_reader.py:44]: Here `npt.assert_equal()` is being used to check lengths of lists. Compare the tracebacks:
```
============================================== FAILURES ===============================================
__________________________________________ test_example_good __________________________________________

    def test_example_good():
>       assert len(list1) == len(list2)
E       assert 2 == 3
E        +  where 2 = len([1, 2])
E        +  and   3 = len([1, 2, 3])

example.py:8: AssertionError
__________________________________________ test_example_bad ___________________________________________

    def test_example_bad():
>       np.testing.assert_equal(len(list1), len(list2))
E       AssertionError: 
E       Items are not equal:
E        ACTUAL: 2
E        DESIRED: 3

example.py:11: AssertionError
```

File: astropy/io/ascii/tests/test_c_reader.py
Comment (eerovaher) [inline astropy/io/ascii/tests/test_c_reader.py:45]: Here `npt.assert_equal()` is being used to check columns of tables. Compare the tracebacks:
```
============================================== FAILURES ===============================================
__________________________________________ test_example_good __________________________________________

    def test_example_good():
>       assert t1.colnames == t2.colnames
E       AssertionError: assert ['a', 'b', 'c'] == ['a', 'c']
E         At index 1 diff: 'b' != 'c'
E         Left contains one more item: 'c'
E         Use -v to get more diff

example.py:11: AssertionError
__________________________________________ test_example_bad ___________________________________________

    def test_example_bad():
>       np.testing.assert_equal(t1.colnames, t2.colnames)
E       AssertionError: 
E       Items are not equal:
E        ACTUAL: 3
E        DESIRED: 2

example.py:14: AssertionError
```

**Review by eerovaher — COMMENTED** [review]
I've commented two uses of `np.testing.assert_equal()`, but analogous comments could be made throughout.

**Comment by dhomeier** [comment]
Those are scalar/string comparisons. I did not expect the older helper functions would even have come into play there, but it looks like originally they were substitutes for unittest functions. I agree in those cases it makes no sense to bring `numpy.testing` in at all, so they will all have to be reviewed case by case.

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_c_reader.py:47]: This is a bit trickier, as meta is neither a sequence nor a scalar, but typically a dict-like structure.
But here the numpy function should still provide better diagnostics:
```python
>>> assert t1.meta == t2.meta
AssertionError
>>> assert_equal(t1.meta, t2.meta)
AssertionError: 
Items are not equal:
key='value'
key='CLIPRANGE'
key='keywords'

 ACTUAL: '2.5'
 DESIRED: '2.0'
```

**Review by dhomeier — CHANGES_REQUESTED** [review]
Actually it seems the majority of `equal`s are just comparing scalars or simple strings; so it might be better to default to `assert` first and then search for array-like and structured objects where to better stick with `assert_equal`.

File: astropy/io/ascii/tests/test_c_reader.py
Comment (eerovaher) [reply astropy/io/ascii/tests/test_c_reader.py:47]: The tracebacks the CPython interpreter produces can be very different from the tracebacks `pytest` produces:
```
============================================== FAILURES ===============================================
__________________________________________ test_example_good __________________________________________

    def test_example_good():
>       assert t1.meta == t2.meta
E       AssertionError: assert {'CLIPRANGE':... 'value': 2.5} == {'CLIPRANGE':... 'value': 2.0}
E         Omitting 2 identical items, use -vv to show
E         Differing items:
E         {'value': 2.5} != {'value': 2.0}
E         Use -v to get more diff

example.py:11: AssertionError
__________________________________________ test_example_bad ___________________________________________

    def test_example_bad():
>       np.testing.assert_equal(t1.meta, t2.meta)
E       AssertionError: 
E       Items are not equal:
E       key='value'
E       
E        ACTUAL: 2.5
E        DESIRED: 2.0

example.py:15: AssertionError

```

Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:47]: Yes, the pytest version might be a bit more useful here still. Would you suggest to leave numpy's `assert_equal` to array-like objects only and use `assert ...==` for everything else?

Comment (eerovaher) [reply astropy/io/ascii/tests/test_c_reader.py:47]: When in doubt, make the test fail and see what the `pytest` traceback looks like. But for this pull request the reasonable thing to do would be to replace `assert_equal()` from `commons.py` with `assert ... == ...`, not `np.testing.assert_equal()`. `commons.assert_equal(x, y)` is basically an alias for `assert x == y` anyways, so that replacement will not make any potential tracebacks worse.

Comment (Harsh-pandya) [reply astropy/io/ascii/tests/test_c_reader.py:47]: @dhomeier  Should I start working on replacing the `npt.assert_equal` back to `assert x == y`.

Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:47]: We have discussed in the meantime that `npt` is not really a standard alias used by astropy, so 
1. those imports should be changed to `numpy.testing import assert_allclose, assert_equal` in general
2. for comparing anything that is not an _array-like collection of numerical values_ (i.e. basically anything that can be converted to a standard numerical-dtype `ndarray`), per the suggestions above, replace `assert_equal` with `assert x == y`.
Thanks!

Comment (Harsh-pandya) [reply astropy/io/ascii/tests/test_c_reader.py:47]: Sure, will start my work as mentioned

Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:47]: Just to clarify, we want
```python
assert_equal(arr_x, arr_y)
assert obj_x == obj_y
```

File: astropy/io/ascii/tests/test_c_reader.py
Comment (eerovaher) [inline astropy/io/ascii/tests/test_c_reader.py:1446]: Why is this being changed?

File: astropy/io/ascii/tests/test_read.py
Comment (eerovaher) [inline astropy/io/ascii/tests/test_read.py:544]: In case of an error we would get a better traceback with [`numpy.testing.assert_array_equal()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html#numpy.testing.assert_array_equal), but the current patch would not make the traceback any worse than it already is. I'll therefore leave it to the sub-package maintainers to decide if the checks should be further modified in this pull request or if there should be a separate follow-up.

I'm highlighting this only here, but there are more analogous checks elsewhere.

File: astropy/io/ascii/tests/test_c_reader.py
Comment (eerovaher) [inline astropy/io/ascii/tests/test_c_reader.py:1309]: Why is this being edited?

Comment (Harsh-pandya) [reply astropy/io/ascii/tests/test_c_reader.py:1309]: Thank you for the review @eerovaher, This needs to be reverted. I have mistakenly made these formatting changes. I will revert this back.

Comment (Harsh-pandya) [reply astropy/io/ascii/tests/test_c_reader.py]: Resolved

File: astropy/io/ascii/tests/test_c_reader.py
Comment (Harsh-pandya) [reply astropy/io/ascii/tests/test_c_reader.py]: Resolved

**Comment by Harsh-pandya** [comment]
Hello, @dhomeier  a gentle reminder for a review. Thank you for your time.

File: astropy/io/ascii/tests/test_read.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:219]:
```suggestion
            assert table.dtype.names == testfile["cols"]
```
These are tuples of strings, so better compared with the standard `assert`

File: astropy/io/ascii/tests/test_read.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:249]:
```suggestion
            assert table.dtype.names == testfile["cols"]
```
as above

File: astropy/io/ascii/tests/test_read.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:544]:
```suggestion
    assert_equal(data["a"].mask, [False, True])
    assert_equal(data["a"], [1, 1])
    assert_equal(data["b"].mask, [False, True])
    assert_equal(data["b"], [2, 1])
```
I agree the `numpy.testing` function provides rather better diagnostics here even though the arrays are small.
This applies to all the `assert (...).all()` checks _here and below_.
In contrast, all the `assert_equal` uses _above this line_ (basically all comparisons of some `dtype.names` are comparing tuples of non-numeric objects and should use `assert ==` – not just those in the suggestions.

File: astropy/io/ascii/tests/test_rst.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_rst.py:30]:
```suggestion
    assert dat.colnames == ["Col1", "Col2"]
```
Here as well, all the "names" comparisons are better done this way

File: astropy/io/ascii/tests/test_c_reader.py
Comment (neutrinoceros) [inline astropy/io/ascii/tests/test_c_reader.py:15]: This looks half-finished. As you saw, we can just inline `assert_equal` instead of replacing it with `numpy.testing.assert_equal`. A couple call sites seem to remain, let's smoke them out !
```suggestion
from numpy.testing import assert_allclose
```

File: astropy/io/ascii/tests/test_c_reader.py
Comment (neutrinoceros) [inline astropy/io/ascii/tests/test_c_reader.py:45]:
```suggestion
    assert t1.colnames == t2.colnames
```

Comment (neutrinoceros) [reply astropy/io/ascii/tests/test_c_reader.py:45]: there's another one on line 1136

File: astropy/io/ascii/tests/test_read.py
Comment (neutrinoceros) [inline astropy/io/ascii/tests/test_read.py:13]: There are still 16 calls to assert_equal in this file to clean up (some of them already pointed out by @dhomeier)
```suggestion
from numpy.testing import assert_allclose
```

File: astropy/io/ascii/tests/test_rst.py
Comment (neutrinoceros) [inline astropy/io/ascii/tests/test_rst.py:6]: 7 occurrences left to clean up here
```suggestion
from numpy.testing import assert_allclose
```

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:15]: > This looks half-finished. As you saw, we can just inline `assert_equal` instead of replacing it with `numpy.testing.assert_equal`. A couple call sites seem to remain, let's smoke them out !

@neutrinoceros there has been an ongoing discussion from
https://github.com/astropy/astropy/pull/15839#issuecomment-1885134788, https://github.com/astropy/astropy/pull/15839#issuecomment-1892419145 and following about when to better use simple `assert` vs. Numpy's `assert_equal` on array-like operands. I am afraid there is no uncontroversial golden path where to use what, but my last set of comments and suggestion was trying to find a reasonable balance of using both options. Do you think `assert x == y` should be always preferred even on numerical ndarrays?

Comment (neutrinoceros) [reply astropy/io/ascii/tests/test_c_reader.py:15]: Oh, sorry I missed it !

> Do you think assert x == y should be always preferred even on numerical ndarrays?

no ! I was trying to be strictly conservative in my review, but I agree that this pattern is best avoided for ndarrays !

**Comment by dhomeier** [comment]
@Harsh-pandya are you still available to finish this effort?

**Comment by Harsh-pandya** [comment]
Hello @dhomeier , Sorry for ghosting on this PR. I won't be able to continue on this PR. Thank you for your time and feedback on the PR.

**Comment by pllim** [comment]
There are already conflicts. And https://github.com/astropy/astropy/pull/16644 is not going to make the conflicts smaller. So closing without merge. Thanks anyway!

**Comment by dhomeier** [comment]
Thank you still for your contribution, @Harsh-pandya, it has sparked a fruitful discussion!

**Pull request closed without merging**
